### ISSUE TITLE
`addProtocol` for custom Protocols like PMTiles

### ISSUE THREAD

**Issue by punov** [maplibre/maplibre-react-native#28]
## Motivation

At the moment, we have a frontend integration with Protomaps vector tiles, that have a full support of Maplibre GL:
https://protomaps.com/docs/frontends/maplibre

Inside our Mobile application, we use Maplibre as tool for rendering maps, but we don't have a chance to render same vector tiles since this custom protocol is not supported out of the box, and there's no way to add it.

### Implementation

Expectations is to have the same `addProtocol` method to provide a loadFn for custom tile servers like pmtiles.
```
import * as pmtiles from "pmtiles";

let protocol = new pmtiles.Protocol();
maplibregl.addProtocol("pmtiles",protocol.tile);
```

If there's any known workaround to provide support for custom protocol using the existing Maplibre API, 
I'd appreciate any directions / advices.

**Comment by jthiard** [comment]
Hi @punov 
I didn't got deep into it, but if I understand it right the `addProtocol` method in maplibre-gl-js is used to override with javascript code the tiles fetching when using a specific protocol in source url.

As this `maplibre-react-native` lib is a wrapper on top of [maplibre-gl-native](https://github.com/maplibre/maplibre-gl-native) - and not maplibre-gl-js, fetching the tiles is made at the native level, and you won't be able to customize it with javascript code.

I'm unsure if it may be customized at the android / iOS level (with java/objectiveC or kotlin/swift hooks) or if it requires to go deeper at the C++ level.

**Comment by punov** [comment]
@jthiard thanks for your answer, after going through the source code, I have the same understanding.
I'm ok to customize the native code to support it, but at the moment I struggle finding the proper hook for tileserver fetch operations.

addProtocol method is as simple as calling the function to fetch the data according to current coordinates and zoom level,
so getting closer to the function that executes it in native code is a step #1 for me.

**Comment by lseelenbinder** [comment]
@punov You'll be looking for functionality around here:

https://github.com/maplibre/maplibre-gl-native/tree/main/src/mbgl/storage

Most likely an adaptation of https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/storage/http_file_source.hpp and https://github.com/maplibre/maplibre-gl-native/blob/main/platform/default/src/mbgl/storage/http_file_source.cpp with some of https://github.com/maplibre/maplibre-gl-native/blob/main/platform/default/src/mbgl/storage/file_source_manager.cpp.

It's a bit of a mess, but generally you can register handlers for types of files, and there's a default FileSourceManager normally invoked. You'll need to create your own and pass it through via ResourceOptions: https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/map/map.cpp#L29.

Does this help point you in the right directly?

**Comment by lpatrun** [comment]
Is there any progress on this? Or some instructions on how to do it?

**Comment by KiwiKilian** [comment]
It's currently beeing worked at the MapLibre Native side https://github.com/maplibre/maplibre-native/pull/2882.

**Comment by lpatrun** [comment]
Is there any plan in the near/distant future to have an addProtocol functionality like it is implemented for the "maplibre-gl"?

**Comment by KiwiKilian** [comment]
AFAIK there is nobody working on `addProtocol`, you can try implementing something like this as described by https://github.com/maplibre/maplibre-react-native/issues/28#issuecomment-1419505405 – but I can't yet imagine a clean way where this would be supported by MapLibre RN.

PMTiles support is much more likely, once it's available in MapLibre Native.

**Comment by tyrauber** [comment]
Wouldn't addProtocol be required to support an api_key on styleJSON?

What about the [maplibre-contour plugin](https://github.com/onthegomap/maplibre-contour) which uses addProtocol under the covers to accomplish the terrain layers from a raster-dem?

**Comment by onizukanaim** [comment]
My solution meanwhile :
Thank you use dom 

```tsx
"use dom";

import maplibregl from "maplibre-gl";
import { Protocol } from "pmtiles";
import layers from "protomaps-themes-base";
import { useEffect } from "react";
import { Map } from "react-map-gl";
import { View } from "react-native";
import { useColorScheme } from "~/utils/useColorScheme";

export default function MapViewWeb() {
  const { colorScheme } = useColorScheme();
  useEffect(() => {
    let protocol = new Protocol();
    maplibregl.addProtocol("pmtiles", protocol.tile);
    return () => {
      maplibregl.removeProtocol("pmtiles");
    };
  }, []);

  return (
    <Map
      attributionControl={false}
      style={{ width: "100vw", height: "100vh" }}
      mapStyle={{
        glyphs:
          "https://protomaps.github.io/basemaps-assets/fonts/{fontstack}/{range}.pbf",
        sprite:
          "https://api.protomaps.com/styles/v4/light/en.json?key=NON_COMMERCIAL_USE_KEY",
        version: 8,
        sources: {
          sample: {
            type: "vector",
            url: "pmtiles://https://world.pmtiles",
          },
        },
        layers: layers(
          "sample",
          colorScheme,
          "en"
        ) as mapboxgl.LayerSpecification[],
      }}
      mapLib={maplibregl as any}
    />
  );
}

```

**Comment by conbrad** [comment]
PMTiles support was recently merged in `maplibre-native`: https://github.com/maplibre/maplibre-native/pull/2882

I think they just cut a release, can we bump that dependency here?

**Comment by KiwiKilian** [comment]
Sure, it's on my radar, but please have a little patience – It's not even yet released/deployed on both platforms.

Once #589 is merged, anyone can control the underlaying MapLibre Native versions quite easily. Which would allow anyone to test a new native release, once it's available, without changes to this repository.

**Comment by KiwiKilian** [comment]
I still think adding `addProtocol` is out of scope for this project, it belongs towards MapLibre Native. And MapLibre Native seems to solve these requirement differently, as we see `pmtiles://` getting an integrated solution.

Quoting @tyrauber 

> Wouldn't addProtocol be required to support an api_key on styleJSON?
> 
> What about the [maplibre-contour plugin](https://github.com/onthegomap/maplibre-contour) which uses addProtocol under the covers to accomplish the terrain layers from a raster-dem?

An `apiKey` should just be added to the URLs as parameters, see [Stadia Maps docs](https://docs.stadiamaps.com/native-multiplatform/maplibre-gl-native/#quickstart) for example.

For contour lines there is an [open design proposal in the spec](https://github.com/maplibre/maplibre-style-spec/issues/583).

So basically, this is mainly always a requirement towards MapLibre Native. If there is ever some `addProtocol` for MapLibre Native, we will for sure try to expose this! Therefore I'm planning to close this issue, once we have PMTiles support through bumping the native dependencies. Feel free to comment, if you see other use cases, but otherwise I recommend to research your use case and propose such solutions towards MapLibre Native, so we can implement on top of it.

**Comment by tyrauber** [comment]
Totally agree, @KiwiKilian, I have no problem letting the Native team do the heavy lifting.

I believe the primary issue with the `apiKey` is that some map tile styles are hosted by a third-party, and adding a apiKey to the style json url doesn't automatically add it to the tile or asset requests.  Perhaps we should create an issue and track that separately.

I would very much like to see [Michael's contour lines proposal](https://github.com/maplibre/maplibre-style-spec/issues/583) get implemented. His JS plugin which I think implements this is very nice.

**Comment by KiwiKilian** [comment]
I'm closing this one, as PMTiles support is available.

Supporting API keys already has it's own issue #424.

If anyone has another use case for `addProtocol` feel free to chime in, or go forward towards MapLibre Native.

**Comment by conbrad** [comment]
> > I'm closing this one, as PMTiles support is available.
> > Supporting API keys already has it's own issue [#424](https://github.com/maplibre/maplibre-react-native/issues/424).
> > If anyone has another use case for `addProtocol` feel free to chime in, or go forward towards MapLibre Native.
> 
> I was waiting for supporting PMTiles protocol in maplibre-rn, is it available in the latest beta release, if so, is there a guide for using it ?

The new versions of `maplibre-native` that contains PMTiles support were bumped and merged here in the beta branch here it looks like (awesome, thank you!): https://github.com/maplibre/maplibre-react-native/releases/tag/v10.0.0-beta.19

You can reference the `maplibre-native` docs to use PMTiles: https://maplibre.org/maplibre-native/android/examples/data/PMTiles/

**Comment by abdelhameedhamdy** [comment]
> > > I'm closing this one, as PMTiles support is available.
> > > Supporting API keys already has it's own issue [#424](https://github.com/maplibre/maplibre-react-native/issues/424).
> > > If anyone has another use case for `addProtocol` feel free to chime in, or go forward towards MapLibre Native.
> > 
> > 
> > I was waiting for supporting PMTiles protocol in maplibre-rn, is it available in the latest beta release, if so, is there a guide for using it ?
> 
> The new versions of `maplibre-native` that contains PMTiles support were bumped and merged here in the beta branch here it looks like (awesome, thank you!): https://github.com/maplibre/maplibre-react-native/releases/tag/v10.0.0-beta.19
> 
> You can reference the `maplibre-native` docs to use PMTiles: https://maplibre.org/maplibre-native/android/examples/data/PMTiles/

Aha, thank you for clarifying that, I thought it was exposed in v10.0.0-beta.19.
